### ISSUE TITLE
add filter alelleles

### ISSUE THREAD

**Issue by cseed** [hail-is/hail#551]
Create a filter alleles command.  The interface should be modeled off the current filter commands (variant, sample, genotype).  One wrinkle is that to update the variant annotations, which must be done by the user, they must have access to information about what alleles were filtered.  Rough sketch of the interface:

```
... filteralleles
  --keep
  -c 'va.alleleQuality[aIndex] > 0.8'
  -a 'va.info.AC = aIndices.map(i => va.info.AC[i]), va.info.AN = ...'
```

where `va.alleleQuality` is a hypothetical annotation of type `Array[Double]`.  It also has a `--remove` option.

`-c` is the filter condition.  It has type `Boolean` and `v`, `va`, and `aIndex` in scope, where `aIndex` is the index of the allele being evaluated.

`-a` is an annotation expression which updates variant annotations analogous to the `-c` argument of `annotatevariants expr`.  It has `v`, `va` and `aIndices`, where `v` is the _new_ variant (so `v.altAlleles.size` is the number of alleles being kept), `va` is the old variant annotations which are being updated, and `aIndices` is a map from the new to old allele indices.

Only alternate alleles should be filtered, although all indices should be 0-based, counting the reference.  If no alternate alleles remain, the variant should be filtered (no monomorphic variants).

Step one should be to sketch the command docs so we can get feedback on the interface.  We should work with @konradjk and Monkol to make examples that handle the ExACv2 use case.

**Comment by cseed** [comment]
Tagging Monkol: @monkollek

**Comment by danking** [comment]
This is primarily a condensation and formatting of what you have above. Is there more to document?


### `filteralleles`

#### Usage
- `-c | --condition <expr>`—a hail language expression

| Name | Description |
| --- | --- |
| `v` | variant |
| `va` | variant annotations |
| `aIndex` | allele index |

- `--keep/--remove`—keep or remove the allele if the expression is true

| Name | Description |
| --- | --- |
| `v` | the _new_ variant |
| `va` | the _old_ variant annotations |
| `aIndices` | an array of the old indices (such that `aIndices[new] = old`) |

- `-a | --annotation <expr>`—a hail language expression which may update the variant annotations based on the removed alleles

*NB:* the allele indices are zero indexed and the zeroth index contains the reference. The `condition` expression will be executed for all alleles with index greater than zero.

*NB2:* if all alternate alleles are filtered the entire variant is filtered.

#### Examples

```
... filteralleles
  --keep
  -c 'va.alleleQuality[aIndex] > 0.8'
  -a 'va.info.AC = aIndices.map(oldIndex => va.info.AC[oldIndex]), va.info.AN = ...'
```

**Comment by cseed** [comment]
Two small comments:

 - make it clear the name/description tables are describing the scope of the corresponding expression.  We might need to do this elsewhere in the documentation, too.

 - I'd write `aIndices[newIndex] = oldIndex` just to make it clear you're talking about the indices, and not the alleles.

**Comment by danking** [comment]
During implementation, I noticed some behaviors that were not fully specified.

If an allele is filtered, we must address the genotypes that reference that allele. A genotype consists of five parts.

|Part|Description|Action|
|---|---|---|
|GT|the hard call|if the filtered allele is `a` then `forall b.` `b/a` and `a/b` are converted, respectively, to `b/0` and `0/b`|
|AD|allele depth|the filtered allele's column is eliminated, e.g. filtering allele 1 transforms `[25,5,20]` to `[25,20]`|
|DP|number of informative reads|no change|
|PL|Phred-likelihoods for each allele pair|convert the allele-pair & likelihood pairs (e.g. `(0/1, 10)`) according to the GT rule. This yields a bag of (possibly duplicated) allele-likelihood pairs. We reduce back to unique allele-pairs by taking the `min` likelihood for each allele-pair|
|GQ|genotype quality|set to the second lowest value in the modified PL|

I'm a tad uneasy about the actions for AD, DP, and PL. For AD, should we shift the depth to the reference? For DP, should we subtract the removed depths? For PL, `min` should be ok when the values have differing orders of magnitude, but if two values are similar, should we convert to probabilities, sum, and convert back?

**Comment by danking** [comment]
Consider three alleles.

```
GT: 1/2
GQ: 10

0 | 1000
1 | 1000   10
2 | 1000   0     20
  +-----------------
    0      1     2
```

Want to preserve:

 - PL(GT) = 0

### Algorithms

Suppose we remove allele 2. There are two options:

 - (*minning*) in the PL array convert occurences of 2 to 0 ("downcode to ref") and take minimums where there are multiple likelihoods for a single genotype; also downcode GT to ref

 - (*subsetting*) subset the PL array (i.e. remove entries with 2's); set GT to the genotype with the minimum likelihood


### Interpretations

The qualitative interpretation of *minning* is a belief that the alternate is real but we want to shift the probability mass to 0 (thus changing our interpretation of 0 from "reference" to "reference or something not listed")

The qualitative interpretation of *subsetting* is a belief that the alternate is not-real and we want to discard any probability mass associated with the alternate.

### Results

The *minning* algorithm produces

```
GT: 0/1
GQ: 10

0 | 20
1 | 0    10
  +-----------
    0    1
```

The *subsetting* algorithm produces

```
GT: 1/1
GQ: 990

0 | 990
1 | 990  0
  +----------
    0    1
```

**Comment by danking** [comment]
Here's an update to the action table

|Part|Description|Action|
|---|---|---|
|GT|the hard call|*minning* or *subsetting*|
|AD|allele depth|the filtered allele's column is eliminated, e.g. filtering allele 1 transforms `[25,5,20]` to `[25,20]`|
|DP|number of informative reads|no change|
|PL|Phred-likelihoods for each allele pair|*minning* or *subsetting*|
|GQ|genotype quality|increasing-sort PL and take `PL[1] - PL[0]`|

We choose either *minning* or *subsetting* consistently for all parts.

I now feel:

 - when *minning* (i.e. *believe real*) we should move AD value to reference.

 - when *subsetting* (i.e. *believe not-real*) we should subtract removed depth from DP

**Comment by danking** [comment]
Final action table based on discussion with @cseed. We decided that DP means (and means nothing more): "number of reads the sequencer saw at this position".

|Part|Description|Action|
|---|---|---|
|GT|the hard call|*minning* or *subsetting*|
|AD|allele depth|*minning* or *subsetting*|
|DP|number of informative reads|no change|
|PL|Phred-likelihoods for each allele pair|*minning* or *subsetting*|
|GQ|genotype quality|increasing-sort PL and take `PL[1] - PL[0]`|

We choose either *minning* or *subsetting* consistently for all parts.

For AD,

 - (*minning*) the filtered allele's column is eliminated and the value is added to the reference, e.g. filtering allele 1 transforms `[25,5,20]` to `[30,20]`

 - (*subsetting*) the filtered allele's column is eliminated, e.g. filtering allele 1 transforms `[25,5,20]` to `[25,20]`

**Comment by konradjk** [comment]
I mostly agree with the last idea, though I would suggest defaulting to (or really only implementing) subsetting rather than minning: minning is highly inconsistent with what's done by GATK. I don't think it's particularly valid to take reads assigned to another haplotype and call them reference (ESPECIALLY for GT - this is likely to lead to some pretty gnarly reference bias that most will want to avoid). And to confirm, what does subset mean in terms of GT? Say for `1/2` where 2 is filtered - will it be `./1`?

**Comment by danking** [comment]
When subsetting, the GT will be set to whatever genotype has the lowest negative-log-likelihood in the PL after subsetting. This maintains the `PL(GT) = 0` invariant, though I could also see an argument for `./.`. AFAIK, we cannot represent half-NA cases like `./1`.

@cseed, we had planned to implement *minning*, should we not do that?

**Comment by konradjk** [comment]
Cool, I'm good with that strategy for PL. Next lowest PL is certainly consistent with GATK behavior.

And maybe minning is ok, but it feels weird given how (at least) I think about sequencing data. Might be worth a discussion with DSDE folks.

**Comment by danking** [comment]
Resolved by PR #658
